mqtt/mqtt-bridge/src/persist/storage/ring_buffer/mod.rs

```diff
@@ -0,0 +1,693 @@
+            }
+            let data_size = block.inner().data_size();
+            let index = block.inner().write_index();
+            start = start + block_size;
+            let end = start + data_size;
+            let bytes =
+                read_data_maybe_wrap_arond(self.inner.clone(), start, end, self.max_file_size);
+            let data = binary_deserialize::<Data>(&bytes)?;
+            start = end % self.max_file_size;
+
+            // to move to beginning of next block header
+            // start += 1;
+            validate(&block, &data)?;
+            let publication = binary_deserialize::<Publication>(data.inner())?;
+            let key = Key {
+                offset: index as u64,
+            };
+            vdata.push_back((key, publication));
+        }
+
+        Ok(vdata)
+    }
+
+    fn remove(&mut self, key: usize) -> StorageResult<()> {
+        let timer = Instant::now();
+        let read_index = self.pointers.read.load(Ordering::SeqCst);
+        if key != read_index {
+            return Err(StorageError::RingBuffer(RingBufferError::RemovalIndex));
+        }
+        let block_size = serialized_block_size()?;
+
+        let start = key;
+        let mut end = start + block_size;
+        if end > self.max_file_size {
+            return Err(StorageError::RingBuffer(RingBufferError::WrapAround.into()));
+        }
+
+        let maybe_block = deserialize_block(self.inner.clone(), start, end);
+        if maybe_block.is_err() {
+            return Err(StorageError::RingBuffer(RingBufferError::NonExistantKey));
+        }
+
+        let data_size;
+        let mut block = maybe_block.unwrap();
+        if block.inner().hint() != BLOCK_HINT {
+            return Err(StorageError::RingBuffer(RingBufferError::NonExistantKey));
+        }
+
+        {
+            let inner_block = block.inner_mut();
+            inner_block.set_should_not_overwrite(false);
+            data_size = inner_block.data_size();
+        }
+
+        let bytes = binary_serialize(&block)?;
+        let should_flush = should_flush(self.flush_options, &self.flush_state);
+        mmap_write(self.inner.clone(), start, end, &bytes, should_flush)?;
+
+        end += data_size;
+
+        self.pointers
+            .read
+            .store(end % self.max_file_size, Ordering::SeqCst);
+
+        self.flush_state
+            .update(0, 0, timer.elapsed().as_millis() as usize);
+        if should_flush {
+            self.flush_state.reset(&self.flush_options);
+        }
+
+        Ok(())
+    }
+
+    fn set_waker(&mut self, waker: &Waker) {
+        self.waker = Some(waker.clone());
+    }
+
+    fn wake_up_task(&mut self) {
+        if let Some(waker) = self.waker.take() {
+            waker.wake();
+        }
+    }
+}
+
+#[allow(dead_code)]
+impl Drop for RingBuffer {
+    fn drop(&mut self) {
+        let mmap = self.inner.lock();
+        mmap.flush().expect("Failed to flush");
```

Comment:
do we really want to panic here? I'm not saying this is bad, just curious.

Reply:
The way I am looking at it is that the RingBuffer exists for the lifetime of the bridge. So, dropping is something that should only happen in rare cases where something else panicked. That is why I added this custom Drop to try and flush at the very last moment if a chance exists, and if we cannot, well we are already panicking so what is one more :)

Reply:
I hope nothing in the bridge should panic :) and panic inside a panic will cause app to terminate immediately, which is the worst case for us. If we can't flush, we already potentially loosing data. Abrupt termination can cause other bridges to loose data. So I would rather prefer ignoring (and maybe logging) the error. Not sure how logging would work here though... @dmolokanov, any thoughts?

Reply:
Good point on immediate termination. I would expect logging to not be guaranteed if we are terminating the process either... So, I can also see just swallowing the error as maybe the best choice...

Reply:
Yep, I agree. Let's try to "safely" clear resources and log an error with the highest level.

Reply:
Good plan will switch to putting an appropriate log message